src/components/include/crt_blkpt.h

```diff
@@ -0,0 +1,298 @@
+	CRT_BLKPT_CRIT_SECT = 2,	/* is only one thread ever going to trigger at a time? */
+} crt_blkpt_flags_t;
+
+#define CRT_BLKPT_EPOCH_BLKED_BITS (sizeof(sched_blkpt_epoch_t) * 8)
+#define CRT_BLKPT_BLKED_MASK       (1 << (CRT_BLKPT_EPOCH_BLKED_BITS - 2))
+#define CRT_BLKPT_BLKED(e)         ((e) &  CRT_BLKPT_BLKED_MASK)
+#define CRT_BLKPT_EPOCH(e)         ((e) & ~CRT_BLKPT_BLKED_MASK)
+
+/* Return != 0 on failure: no ids to allocate */
+static inline int
+crt_blkpt_init(struct crt_blkpt *blkpt)
+{
+	sched_blkpt_id_t id;
+
+	id = sched_blkpt_alloc();
+	if (id == SCHED_BLKPT_NULL) return -1;
+
+	*blkpt = (struct crt_blkpt){
+		.id = id,
+		.epoch_blocked = 0
+	};
+
+	return 0;
+}
+
+static inline int
+crt_blkpt_teardown(struct crt_blkpt *blkpt)
+{
+	return sched_blkpt_free(blkpt->id);
+}
+
+/* Internal APIs that must be inlined to remove the branches */
+static inline int
+__crt_blkpt_atomic_trigger(sched_blkpt_epoch_t *ec, sched_blkpt_epoch_t chkpt, crt_blkpt_flags_t flags)
+{
+	/*
+	 * Assume that the most significant bit is the blocked
+	 * indicator. This math might reset it to zero, which we want
+	 * to do anyway (as part of CRT_BLKPT_EPOCH).
+	 */
+	sched_blkpt_epoch_t new = CRT_BLKPT_EPOCH(chkpt + 1);
+
+	/* inlined so that constant propagation will get rid of condition */
+	if (flags == CRT_BLKPT_UNIPROC) {
+		return ps_upcas(ec, chkpt, new);
+	} else {
+		return ps_cas(ec, chkpt, new);
+	}
+	/* TODO: faa for CRT_BLKPT_CRIT_SECT? */
+}
+
+/*
+ * If we return 1, then the caller will attempt to block, otherwise,
+ * return 0 and it will re-check the data-structure assuming that
+ * something happened in the mean time.
+ */
+static inline int
+__crt_blkpt_atomic_wait(sched_blkpt_epoch_t *ec, sched_blkpt_epoch_t chkpt, crt_blkpt_flags_t flags)
+{
+	sched_blkpt_epoch_t cached = ps_load(ec);
+	sched_blkpt_epoch_t new    = cached | CRT_BLKPT_BLKED_MASK;
+	int ret;
+
+	/*
+	 * We are the second or later blocker. Blocked already
+	 * set. We're done here.
+	 *
+	 * It isn't clear if it is better to have the additional
+	 * branch here for this to avoid atomic instructions, or to
+	 * just always do the atomic instructions and possibly fail.
+	 */
+	if (cached == new) return 1;
+
+	/* function is inlined so that constant propagation will get rid of condition */
+	if (flags == CRT_BLKPT_UNIPROC) {
+		ret = ps_upcas(ec, chkpt, new);
+	} else {
+		ret = ps_cas(ec, chkpt, new);
+	}
+	if (unlikely(!ret)) {
+		/*
+		 * CAS failure can mean that 1. another thread
+		 * blocked, and set the blocked bit, or 2. an event is
+		 * triggered. In the former case, we still want to
+		 * block. In the latter case, we want to go back to
+		 * the data-structure.
+		 */
+		return ps_load(ec) == new; /* same epoch with blocked set? == success */
+	}
+
+	return 1;
+}
+
+/* Trigger an event, waking blocked threads. */
+static inline void
+crt_blkpt_trigger(struct crt_blkpt *blkpt, crt_blkpt_flags_t flags)
```

Comment:
This is waking up all the threads I assume... Is that correct? So that they will compete again for the blockpoint